servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ScanConcatPublisher.java

```diff
@@ -0,0 +1,197 @@
+         * {@link #demand} underflow in onNext (in case the source doesn't deliver a timely error).
+         */
+        private static final long INVALID_DEMAND = -1;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<ScanConcatSubscriber> demandUpdater =
+                newUpdater(ScanConcatSubscriber.class, "demand");
+        private final Subscriber<? super R> subscriber;
+        private final Subscriber<? super R> offloadedSubscriber;
+        private final ScanConcatMapper<T, R> mapper;
+        private volatile long demand;
+        @Nullable
+        private Throwable errorCause;
+
+        ScanConcatSubscriber(final Subscriber<? super R> subscriber, final Subscriber<? super R> offloadedSubscriber,
+                             final ScanConcatMapper<T, R> mapper) {
+            this.subscriber = subscriber;
+            this.offloadedSubscriber = offloadedSubscriber;
+            this.mapper = requireNonNull(mapper);
+        }
+
+        @Override
+        public void onSubscribe(final PublisherSource.Subscription subscription) {
+            subscriber.onSubscribe(new PublisherSource.Subscription() {
+                @Override
+                public void request(final long n) {
+                    if (!isRequestNValid(n)) {
+                        handleInvalidDemand(n);
+                    } else if (demandUpdater.getAndAccumulate(ScanConcatSubscriber.this, n,
+                            FlowControlUtils::addWithOverflowProtectionIfNotNegative) == TERMINAL_PENDING) {
+                        demand = TERMINATED;
+                        if (errorCause != null) {
+                            deliverOnError(errorCause, offloadedSubscriber);
+                        } else {
+                            deliverOnComplete(offloadedSubscriber);
+                        }
+                    } else {
+                        subscription.request(n);
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    subscription.cancel();
+                }
+
+                private void handleInvalidDemand(final long n) {
+                    // If there is a terminal pending then the upstream source cannot deliver an error because
+                    // duplicate terminal signals are not allowed. otherwise we let upstream deliver the error.
+                    if (demandUpdater.getAndSet(ScanConcatSubscriber.this, INVALID_DEMAND) == TERMINAL_PENDING) {
+                        demand = TERMINATED;
+                        offloadedSubscriber.onError(newExceptionForInvalidRequestN(n));
+                    } else {
+                        subscription.request(n);
+                    }
+                }
+            });
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            demandUpdater.decrementAndGet(this);
```

Comment:
Should we decrement if `mapper.onNext(t)` throws before `subscriber.onNext` is invoked?

Reply:
good point! we should defer the decrement until after the mapper returns.